### PR TITLE
Fix poisoning

### PR DESCRIPTION
**Pull request by asim29**
Refactored the code a bit. For some reason when only non-trigger-label data is used to poison the model, the test accuracy on the original dataset is reduced. Trying to look into this issue. 

Will test it on CelebA and if everything goes well it should be good to go.

### CODE REVIEW

File: amulet/poisoning/attacks/badnets.py [hunk 1]
```diff
@@ -39,22 +39,30 @@ def __init__(
...
-    def reshape(self, data, dataset_name="fmnist"):
-        if dataset_name == "fmnist":
-            new_data = data.unsqueeze(1)
-        elif dataset_name == "cifar10":
-            new_data = np.transpose(data, (0, 3, 1, 2))
+    def __poison_datapoint(self, data_original: torch.Tensor) -> torch.Tensor:
+        data_point = data_original.clone().detach()
+        if self.dataset_type == "image":
+            channels, width, height = data_point.shape
+            for c in range(channels):
```
Comment (sebszyller) [inline amulet/poisoning/attacks/badnets.py:54]: what's exactly supposed to happen here? looks like a black border close to the edge

Comment (sebszyller) [reply amulet/poisoning/attacks/badnets.py:54]: we'd some comments

File: amulet/poisoning/attacks/badnets.py [hunk 2]
```diff
@@ -68,41 +76,30 @@ def poison_dataset(self, dataset, mode="train"):
...
-        perm = np.random.default_rng(seed=self.random_seed).permutation(len(dataset))[
-            0 : int(len(dataset) * self.portion)
-        ]
-
-        # TODO: Figure out a more efficient way of doing this without a for loop
         data_points = []
         targets = []
-        if self.dataset_name == "fmnist" or self.dataset_name == "cifar10":
-            channels, width, height = dataset[0][0].shape
+        if mode == "test":
```
Comment (sebszyller) [inline amulet/poisoning/attacks/badnets.py:81]: could you speak more to it? The description in the PR isn't clear enough to me

**Review by sebszyller — COMMENTED** [review]
I think the content of the changes is fine once I understand their nature

File: amulet/poisoning/attacks/badnets.py (hunk 1 above)
Comment (asim29) [reply amulet/poisoning/attacks/badnets.py:54]: in the original BadNets code they were setting these values to 255. I changed it to 0. I also moved it to a separate function because it's being called twice. I believe in the case of images you are correct, a small black square close to the edge. In the case 1D data it's the last 5th of the feature vector being set to 0.

File: amulet/poisoning/attacks/badnets.py (hunk 2 above)
Comment (asim29) [reply amulet/poisoning/attacks/badnets.py:81]: In the case of a "test" dataset, all the samples are poisoned. In the other case when used for training, a proportion of data points are poisoned. 

Initially, I tried to only poison non-trigger-label samples. However, that results in terrible test accuracy on the un-poisoned dataset. Not exactly sure why.

Comment (sebszyller) [reply amulet/poisoning/attacks/badnets.py:81]: what does it mean that the test is poisoned?
When I read trigger-label, I understand "I poison only records that have class X" which seems a bit weird.

Could you remind what kind of poisoning is badnets doing?

Comment (asim29) [reply amulet/poisoning/attacks/badnets.py:81]: Let's say it's a 2-class problem, where class 1 is the attacker's "goal" label, i.e. the attacker wants all poisoned data points to be classified is 1. 

To evaluate a poisoned model against poisoned data points, we take the test set, and take **only** the data points from class 0, add the backdoor (or trigger), and change the label to 1. On a normal, unpoisoned, model, this should _ideally_ give 0% accuracy (because all the data points are _actually_ of class 0, but the label says they are class 1).  But for a poisoned model, this should _ideally_ give 100% accuracy. 

In the case of creating a poisoned training set (to train a poisoned model), we take a random set of **all** data points, add the backdoor, change the label to 1, and add them back into the training set. You mentioned once in a meeting that ideally, we should take data points that don't belong to class 1, poison them, and add them back into the dataset. But, that ruins the test accuracy on the "normal", unpoisoned dataset. 

From the paper: 
> Specifically, we randomly pick p|Dtrain| from the training dataset, where p ∈ (0, 1], and add backdoored versions of these images to the training dataset. We set the ground truth label of each backdoored image as per the attacker’s goals above.

If this is still confusing we can probably set up a short call or discuss it end-of-the-week?
